app/lib/Emoji.js

```diff
@@ -1,78 +1,49 @@
       string = escape(string);
     }
 
-    // Start with replacing BK emojis (which are more likely than Apple emojis)
-    string = this._replace(BUILDKITE_EMOJIS, COLON_REGEXP, string);
+    // Start with replacing Buildkite emoji with images
+    string = this._replace(string, BUILDKITE_EMOJI, (emoji, { host }) => {
+      // Emoji catalogue hosts have a normalized host that always end with a "/"
+      const emojiUrl = `${host}${emoji.image}`;
 
-    // Then do an Apple emoji parse
-    if (options.replaceUnicode !== false) {
-      string = this._replace(APPLE_EMOJIS, UNICODE_REGEXP, string);
-    }
-    string = this._replace(APPLE_EMOJIS, COLON_REGEXP, string);
-
-    return string;
-  }
-
-  _replace(catalogue, regexp, string) {
-    const matches = string.match(regexp);
-    const replacements = [];
-
-    // Bail if there aren't any emojis to replace
-    if (!matches || !matches.length) {
-      return string;
-    }
-
-    for (let matchIndex = 0, matchLength = matches.length; matchIndex < matchLength; matchIndex++) {
-      const match = matches[matchIndex];
-      const nextMatch = matches[matchIndex + 1];
+      // Prioritise unicode representation over shortcodes
+      const emojiCanonicalRepresentation = emoji.unicode || `:${emoji.name}:`;
 
-      // See if this match and the next one, makes a new emoji. For example,
-      // :fist::skin-tone-4:
-      if (nextMatch) {
-        const modifiedEmojiIndex = catalogue.index[`${match}${nextMatch}`];
+      return `<img class="emoji" title="${emoji.name}" alt="${emojiCanonicalRepresentation}" src="${emojiUrl}" draggable="false" />`;
+    });
 
-        if ((typeof modifiedEmojiIndex) === 'number') {
-          replacements.push(this._image(catalogue, catalogue.emoji[modifiedEmojiIndex]));
-          replacements.push("");
-          matchIndex += 1;
+    // Replace Unicode emoji shortcodes with real Unicode
+    string = this._replace(string, UNICODE_EMOJI, ({ unicode }) => {
+      return unicode;
+    });
 
-          continue;
-        }
-      }
+    return string;
+  }
 
+  _replace(string, catalogue, renderer) {
+    return string.replace(COLON_REGEXP, (match) => {
       const emojiIndex = catalogue.index[match];
 
       if ((typeof emojiIndex) === 'number') {
-        replacements.push(this._image(catalogue, catalogue.emoji[emojiIndex]));
-      } else {
-        replacements.push(match);
+        return renderer(catalogue.emoji[emojiIndex], catalogue);
       }
-    }
-
-    return string.replace(regexp, () => replacements.shift());
-  }
-
-  _image(catalogue, emoji) {
-    // Emoji catalogue hosts have a normalized host that always end with a "/"
-    const emojiUrl = `${catalogue.host}${emoji.image}`;
-    const emojiCanonicalRepresentation = emoji.unicode || `:${emoji.name}:`;
+      return match;
 
-    return `<img class="emoji" title="${emoji.name}" alt="${emojiCanonicalRepresentation}" src="${emojiUrl}" draggable="false" />`;
+    });
```

Comment:
Nice. :100:

One of the best performance tweaks I made to the backend emoji processing code was inverting the loops, changing to a single scan over the text which then looped through the catalogues to find matches. Is that something which might help here? I know they behave slightly differently here but that behaviour could be combined.